docs/design/in-doc-search-ui.rst

```diff
@@ -0,0 +1,183 @@
+* Support across all (or virtually all) Sphinx themes.
+* Support for the JavaScript user experience down to IE11 or graceful degradation where we can't support it.
+* Project maintainers should have a way to opt-in/opt-out of this feature.
+* (Optional) Project maintainers should have the flexibility to change some of the styles using custom CSS and JS files.
+
+Non-Goals
+++++++++++
+
+* For the initial release, we are targeting only Sphinx documentations
+  as we don't index MkDocs documentations to our Elasticsearch index.
+
+
+Existing Search Implementation
+------------------------------
+
+We have a detailed documentation explaing the underlying architecture of our search backend
+and how we index documents to our Elasticsearch index.
+You can read about it :doc:`here <../development/search>`.
+
+
+Proposed Architecture for In-Doc Search UI
+------------------------------------------
+
+Frontend
+++++++++
+
+Technologies
+~~~~~~~~~~~~
+
+Frontend is to designed in a theme agnostics way. For that,
+we explored various libraries which may be of use but none of them fits our needs.
+So, we might be using vanilla JavaScript for this purpose.
+This will provide us some advantages over using any third party library:
+
+* Better control over the DOM.
+* Performance benefits.
+
+
+Proposed Architecture
+~~~~~~~~~~~~~~~~~~~~~
+
+We plan to select the search bar, which is present in every theme,
+and use the `querySelector()`_ method of JavaScript.
+Then add an event listener to it to listen for the changes and
+fire a search query to our backend as soon as there is any change.
+Our backend will then return the suggestions,
+which will be shown to the user in a clean and minimal UI.
+We will be using `document.createElement()`_ and `node.removeChild()`_ method
+provided by JavaScript as we don't want empty `<div>` hanging out in the DOM.
+
+We have a few ways to include the required JavaScript and CSS files in all the projects:
+
+* Add CSS into `readthedocs-doc-embed.css` and JS into `readthedocs-doc-embed.js`
+  and it will get included.
+* Package the in-doc search into it's own self-contained CSS and JS files
+  and include them in a similar manner to `readthedocs-doc-embed.*`.
+* It might be possible to package up the in-doc CSS/JS as a sphinx extension.
+  This might be nice because then it's easy to enable it on a per-project basis.
+  When we are ready to roll it out to a wider audience,
+  we can make a decision to just turn it on for everybody (put it in `here`_)
+  or we could enable it as an opt-in feature like the `404 extension`_.
+
+
+UI/UX
+~~~~~
+
+We have two ways which can be used to show suggestions to the user.
+
+* Show suggestions below the search bar.
+* Open a full page search interface when the user click on search field.
+
+
+Backend
++++++++
+
+We have a few options to support `search as you type` feature,
+but we need to decide that which option would be best for our use-case.
```

Comment:
Using one backend or the other does involve a lot of changes or it's just changing a setting and a class or similar? I want to know if this is something that we can test and see how it perform and decide after that or if testing this way is complicated since they are two different implementations.

Reply:
If we want to have results like [comment #1](https://github.com/rtfd/readthedocs.org/pull/5707#issuecomment-493784610) then we have to do some changes in the backend and a new API endpoint. The changes may increase the size of elasticsearch index to x3 or x5. (In the demo, I did changes with my elasticsearch index locally to produce the results.)
Further changes can be.. like if we want to support section linking. Currently we just open the result page and user have to scroll down to the required section. For that... there will be considerable big changes. many parts of the elasticsearch will change -- how we index data/search data.

Without any changes is okay too...the test-builds - https://readthedocs.org/projects/test-builds-dojutsu-user/versions/ are without any changes.